### PR TITLE
[Parser] Detect non-breaking space (U+00A0) and offer a fix-it

### PR DESCRIPTION
**Pull request by hashemi**
Detect Unicode non-breaking space character (U+00A0) in the lexer, diagnose it, and offer a fix-it turning it into a regular space. Also treat the non-breaking space as whitespace when considering whether an operator is left/right bound.

Resolves [SR-7122](https://bugs.swift.org/browse/SR-7122).

### CODE REVIEW

File: lib/Parse/Lexer.cpp [hunk 1]
```diff
@@ -716,6 +722,12 @@ static bool isRightBound(const char *tokEnd, bool isLeftBound,
     else
       return true;
 
+  case '\302':
+    if (tokEnd[1] == '\240')
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: Could you make these `'\xC2'` and `'\xA0'`?

File: lib/Parse/Lexer.cpp [hunk 2]
```diff
@@ -2426,6 +2438,18 @@ void Lexer::lexTrivia(syntax::Trivia &Pieces, bool IsForTrailingTrivia) {
       break;
     }
     break;
+  case '\302':
+    if (CurPtr[0] == '\240') {
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: ditto

**Review by rintaro — COMMENTED** [review]
This looks great! Just a few code style requests.

**Comment by rintaro** [comment]
Also, could you add tests for fix-it in `test/Parse` ?

**Comment by hashemi** [comment]
Thanks @rintaro, all suggestions incorporated.
I made a new file for the parse tests. Is there a better spot? Recovery.swift is another possibility.

**Comment by rintaro** [comment]
I think dedicated test file is OK. Since this test is handling invisible characters, it's more safe from accidental (automatic) edits.
@swift-ci Please smoke test

File: lib/Parse/Lexer.cpp [hunk 3]
```diff
@@ -716,6 +722,12 @@ static bool isRightBound(const char *tokEnd, bool isLeftBound,
     else
       return true;
 
+  case '\xC2':
+    if (tokEnd[1] == '\xA0')
```
Comment (hashemi) [inline lib/Parse/Lexer.cpp]: Can you confirm that this dereference is safe? Can we assume that the buffer is always null-terminated?

Comment (rintaro) [reply lib/Parse/Lexer.cpp]: Yes, it's guaranteed. https://github.com/apple/swift/blob/9ba05efb/lib/Parse/Lexer.cpp#L188

**Comment by rintaro** [comment]
Thank you Ahmad! Please mark the SR issue RESOLVED.

**Pull request merged**